sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java

```diff
@@ -0,0 +1,349 @@
+
+      case ENUM:
+        return convertEnumStrict((GenericEnumSymbol) value, fieldType);
+
+      case ARRAY:
+        return convertArrayStrict((List<Object>) value, unwrapped.getElementType(), fieldType);
+
+      case MAP:
+        return convertMapStrict(
+            (Map<CharSequence, Object>) value, unwrapped.getValueType(), fieldType);
+
+      case UNION:
+        throw new RuntimeException("Can't convert 'union', only nullable fields are supported");
+
+      case NULL:
+        throw new RuntimeException("Can't convert 'null' to non-nullable field");
+
+      default:
+        throw new AssertionError("Unexpected AVRO Schema.Type: " + unwrapped.getType());
+    }
+  }
+
+  private static org.apache.avro.Schema unwrapNullableSchema(org.apache.avro.Schema avroSchema) {
+    if (avroSchema.getType() == org.apache.avro.Schema.Type.UNION) {
+      org.apache.avro.Schema nullSchema =
+          org.apache.avro.Schema.create(org.apache.avro.Schema.Type.NULL);
+      List<org.apache.avro.Schema> types = avroSchema.getTypes();
+
+      // optional fields in AVRO have form of:
+      // {"name": "foo", "type": ["null", "something"]}
+      if (types.size() == 2 && types.get(0).equals(nullSchema)) {
+        return unwrapNullableSchema(types.get(1));
+      }
+    }
+
+    return avroSchema;
+  }
+
+  private static Object convertRecordStrict(GenericRecord record, Schema.FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case ROW:
+        return toRowStrict(record, fieldType.getRowSchema());
+      default:
+        throw new RuntimeException("Can't convert 'record' to " + fieldType.getTypeName());
+    }
+  }
+
+  private static Object convertBytesStrict(ByteBuffer bb, Schema.FieldType fieldType) {
+    // array of byte shouldn't be there because it's encoded as List
+    switch (fieldType.getTypeName()) {
+      case BYTES:
+        byte[] bytes = new byte[bb.remaining()];
+        bb.get(bytes);
+        return bytes;
+      default:
+        throw new RuntimeException("Can't convert 'fixed' to " + fieldType.getTypeName());
+    }
+  }
+
+  private static Object convertFixedStrict(GenericFixed fixed, Schema.FieldType fieldType) {
+    // array of byte shouldn't be there because it's encoded as List
+    switch (fieldType.getTypeName()) {
+      case BYTES:
+        return fixed.bytes().clone(); // clone because GenericFixed is mutable
+      default:
+        throw new RuntimeException("Can't convert 'fixed' to " + fieldType.getTypeName());
+    }
+  }
+
+  private static Object convertStringStrict(CharSequence value, Schema.FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case STRING:
+        return value.toString();
+      default:
+        throw new RuntimeException("Can't convert 'string' to " + fieldType.getTypeName());
+    }
+  }
+
+  private static Object convertIntStrict(Integer value, Schema.FieldType fieldType) {
+    switch (fieldType.getTypeName()) {
+      case INT32:
```

Comment:
Thanks for the suggestion. True! I've started with doing implicit casting and then removed this code because it was too complicated, now I will change to guards because there is no need in switch anymore.